File: src/main/java/com/secuworm/endpointcollector/domain/EndpointRepository.java
package com.secuworm.endpointcollector.domain;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class EndpointRepository {
    private final Map<String, EndpointRecord> endpointMap = new LinkedHashMap<>();

    public synchronized EndpointRecord upsert(
        String endpointUrl,
        String host,
        String sourceUrl,
        String sourceRequestMethod,
        List<RequestHeader> sourceRequestHeaders
    ) {
        long now = System.currentTimeMillis();
        EndpointRecord record = endpointMap.get(endpointUrl);
        if (record == null) {
            EndpointRecord created = new EndpointRecord(
                endpointUrl,
                host,
                sourceUrl,
                sourceRequestMethod,
                sourceRequestHeaders,
                1,
                now,
                now
            );
            endpointMap.put(endpointUrl, created);
            return created;
        }

        record.incrementCount();
        record.setLastSeenAt(now);
        return record;
    }

    public synchronized EndpointRecord upsert(String endpointUrl, String host, String sourceUrl) {
        return upsert(endpointUrl, host, sourceUrl, "", null);
    }

    public synchronized List<EndpointRecord> getAll() {
        return new ArrayList<>(endpointMap.values());
    }

    public synchronized int size() {
        return endpointMap.size();
    }

    public synchronized void clear() {
        endpointMap.clear();
    }
}

File: src/domain/endpoint_repository.py
import time

from domain.models import EndpointRecord


class EndpointRepository(object):
    def __init__(self):
        self.endpoint_map = {}

    def upsert(self, endpoint_url, host, source_url):
        now_ms = int(time.time() * 1000)
        record = self.endpoint_map.get(endpoint_url)
        if record is None:
            self.endpoint_map[endpoint_url] = EndpointRecord(
                endpoint_url=endpoint_url,
                host=host,
                source_url=source_url,
                count=1,
                first_seen_at=now_ms,
                last_seen_at=now_ms,
            )
            return self.endpoint_map[endpoint_url]

        record.count += 1
        record.last_seen_at = now_ms
        return record

    def get(self, endpoint_url):
        return self.endpoint_map.get(endpoint_url)

    def get_all(self):
        return self.endpoint_map.values()

    def size(self):
        return len(self.endpoint_map)

    def clear(self):
        self.endpoint_map = {}
